Long file names and null-terminated file names

bitcog opened the pull request:
Hi Tomasz,

Thanks for writing and sharing targz-sizes, it's really useful to me.

For a project that I'm working on, I needed to:
  (1) lift the 100 character filename limit of targz-sizes, and
  (2) have null-terminated filenames null-terminated.
In this pull request, I've implemented both the above things.  I've also added some, unit tests, debug macros and built-in Valgrind support.

Feel free to pull these changes - no problem if you don't want them.  I'm happy to consider your review feedback if you want to provide that too.

Regards,

Douglas

bitcog committed:
Added unit tests.

bitcog committed:
Enable Valgrind for unit tests.

bitcog committed:
Added command line arguments and debug macros.

bitcog committed:
Added option for file names to be null-terminated.

bitcog committed:
Added support and unit tests for long file names.

bitcog committed:
Added support and unit tests for long file names.

bitcog committed:
Merge branch 'master' of https://github.com/bitcog/targz-sizes

liori commented on test/test-common.sh:
```diff
@@ -0,0 +1,48 @@
+#!/bin/sh
+# vim: ts=4 sw=4 ts
+
+test_id=$(echo $0 | sed -e 's/^[^a-z]*//' | sed -e 's/.sh$//')
```
Please use shell substitutions instead. For example, "${0%.sh}" will remove the extension. See the section "Parameter Expansion" of e.g. dash for more details (for convenience: http://linux.die.net/man/1/dash).

bitcog replied:
Great feedback.  For years, I've been using sub-processes to do this work, this is much cleaner.  Addressed in commit: 4b3e33ac9eb7c8d072bc9e763424c400539352ef

liori commented on test/test-common.sh:
```diff
@@ -0,0 +1,48 @@
+runTest () {
+
+    # run the test
+    echo "runTest(), run command: ${LOG_COMPILER} ${LOG_FLAGS} ${TEST_COMMAND} < ${test_dir}/input.tgz > ${test_dir}/rawout 2> ${test_dir}/rawerr"
```
I'd extract this pattern of `echo`ing a command, then executing it, then printing its exit code to a separate shell function. You're using it in several places here.

bitcog replied:
Good pickup.  Addressed in commit: b860350a7d411c7fb36618cfa1de4af1dd414d30

liori commented on test/test-common.sh:
```diff
@@ -0,0 +1,48 @@
+	cat ${test_dir}/rawerr ${test_dir}/rawout
+    echo "runTest(), command_result=$command_result"
+    if [ "$command_result" != "0" ] ; then
+        echo "runTest(), exiting"
```
Indentation is pretty inconsistent in this file. Please decide on either tab or 4 spaces.

bitcog replied:
Mmmm, wierd.  I can see on line 21 of the code snippet (above) that the `cat` command is indented further than the other commands that should be at the same level.  But I've searched the file for tab characters (both the latest revision and the revision that you reviewed) and can't find any.  If you can still see this problem, could you please give me the file name, line number and character number where you can see a tab, and I'll take another look.

liori commented:
Well, firstly, thank you for your interest, it's actually the first pull request I received for my project!

I started reviewing your changes the way I usually do at work, with extreme nitpicking and all that stuff, but then realized that maybe it's not worth it for such a short piece of code. So, if you're planning to fix anything basing on my comments, you're free to do so within this branch/pull request. If not, just say so and I'll merge it as is. Of course, if you want the nitpicking I usually do to my coworkers, I can do that too ;-)

bitcog commented:
Thanks for your feedback.  This is the first time I've contributed to a project, so I would like to get all the feedback I can get.  Yes, I would appreciate the nitpicking review feedback if you have the time.  So I'll fix and commit the fixes in the branch/pull request so that it can be later merged.

bitcog commented:
Not having used git before, I've now read about the code review process and discovered that I should have made a new branch before delivering my changes. Do you think I should have a go at moving my commits onto a new branch and make a new pull request for that?

liori commented:
Sorry for not responding earlier, I was pretty busy. Sure, if you want to learn how to do it the github way, go for it. I'm not very familiar with it either. I'll look at your changes again today.

liori commented on test/basic.sh:
```diff
@@ -0,0 +1,20 @@
+#!/bin/sh
+
+. ${top_srcdir}/test/test-common.sh
```
Almost each time you're using variable substitution, you should quote it. Otherwise, if the variable contains spaces (e.g. a path with spaces), it will be expanded into two separate words, not a single argument. In this file alone you should do so in this line, lines 7, 8 (twice) and 16. The usual rule is: "If in doubt, quote." ;-)

bitcog replied:
Addressed in this file and other .sh files, in commit: e5812284788adb558f15f71ca4851cde668ddff6

liori commented on src/targz_sizes.c:
```diff
@@ -10,12 +11,39 @@
 #define TAR_BLOCK_SIZE 512
 
+#ifndef max
+    #define max( a, b ) ( ((a) > (b)) ? (a) : (b) )
```
I'd prefer the `min` and `max` macros to be normal inline static functions; macros were historically used because (1) old compilers didn't know how to inline functions properly (but for at least 15 years now every compiler does that), (2) to make them work for any type with well-defined `<` (but here only `int`s matter).

bitcog replied:
Yes, inline static functions look better.  Addressed in commit: 9a23a4c784eb5fb6828ccdfe9835bd58f67b13da

liori commented on src/targz_sizes.c:
```diff
@@ -120,8 +126,24 @@ int main(int argc, char** argv) {
     }
 
+    unsigned int max_filename_blocks;
+    if ( LEVEL_IS_TEST( verbosity ) ) {
```
Formatting: avoid spaces right after `(` and right before `)` to make it consistent with the rest of this file.

bitcog replied:
I'm sorry that you are having to pick-up formatting inconsistencies.  Addressed in commit: 884ca7e05ad490a629e41738a69fe938cd74682a

liori commented on src/targz_sizes.c:
```diff
@@ -120,8 +126,24 @@ int main(int argc, char** argv) {
+        // when targz_sizes is in *test* mode, set max_filename_blocks
+        // very small (2).  This makes it easier to test the case where
+        // a long filename would overrun the buffer
+        max_filename_blocks=2;
```
Formatting: put a single space before and after the assignment operator to make it consistent with the rest of this file.

bitcog replied:
Again, sorry about the formatting inconsistencies.  Addressed in commit: 311a6eebb9e5cf57e4e9a6d74d70e0f733f905ce

liori commented on src/targz_sizes.c:
```diff
@@ -120,8 +126,24 @@ int main(int argc, char** argv) {
     unsigned char decompressed[TAR_BLOCK_SIZE];
+    tarblock_t *filename_buff;
+    LOG_INFO(verbosity, "sizeof(tarblock_t)=%d", sizeof(tarblock_t) );
+    filename_buff = (tarblock_t*) calloc(
```
In C99 you can make a local array with size given by another variable. Here you can just write `char filename_buff[max_filename_blocks * sizeof(tarblock_t)]`. Then you don't need to remember to `free()` it.

bitcog replied:
OK, my lack of 'C' experience is starting to show through here.  I done my best with this one and it looks good to me.  Addressed in commit: 38ca88b5917f09d66fce33819c7a19d92cd693d2

liori commented on src/targz_sizes.c:
```diff
@@ -137,7 +159,11 @@ int main(int argc, char** argv) {
+    int file_blocks = 0;
     int skip_blocks = 0;
+    int read_blocks = 0;
+    unsigned char* long_filename = (unsigned char*) filename_buff;
```
You're breaking the strict aliasing rule here. For explanation, see e.g. https://stackoverflow.com/questions/98650/what-is-the-strict-aliasing-rule. For this code, just declare the original `filename_buff` as an array of chars with an appropriate size, should be enough for this code.

bitcog replied:
And I've struggled a bit with this one.  Anyway, I've addressed the comment and made other (what seem to be appropriate) corresponding updates. Addressed in commit: 38baeb79a0eb820299e0411fad1cea7b263f96e2

liori commented on README:
```diff
@@ -91,9 +102,3 @@ if you want to fix them.
 Archive file names are not re-encoded to current locale. You can use `iconv`
 for that.
-
-Archive file names are truncated to 100 characters (the limit for original tar
```
As far as I understand, there are 3 incompatible tar extensions that handle long file names. This is one of the reasons I didn't bother to implement long file names in the first place ;-) Anyway, please document here that this code works with GNU-style long file names, but not POSIX-style or old-GNU.

bitcog replied:
This subtlety got past me.  I've restored your original _known problem_ expanding it to describe how it handles the GNU long file name extension.  Addressed in commit: 40b6193be24af863a930163edb612be0b6026d1d

liori commented:
Ok, I think that's it for today. Have fun!

bitcog commented:
Thank you for all the detailed feedback.  I've read through it and it all makes sense.

Sorry it has been a while to respond, I've been busy too, and I didn't have my notifications set up properly, so I didn't notice all your feedback until now.  It might still be a few weeks before I address the feedback.

bitcog committed:
github pull request #6, review comment 1 rework. Shell substitution is now used

bitcog committed:
github pull request #6, review comment 2 rework. Factor out checkCommand function

bitcog committed:
github pull request #6, review comment 3 rework. Add quotes to protect shell variable expanstion from breaking into words

bitcog committed:
github pull request #6, review comment 4 rework. Replace min and max macros with inline static functions

bitcog committed:
github pull request #6, review comment 5 rework

Make round bracket formatting consistent

bitcog committed:
github pull request #6, review comment 6 rework

Make assignment operator formatting consistent

bitcog committed:
github pull request #6, review comment 7 rework

Convert the dynamically allocated array into a local array

bitcog committed:
github pull request #6, review comment 8 rework

Update to comply with the strict aliasing rule

bitcog committed:
github pull request #6, review comment 9 rework

Restore the description of the _known problem_ which relates to the
100-byte file name limitation.  This discription now states that the
GNUTYPE_LONGNAME extension is supported.

bitcog commented:
Well, yes it did take me a few weeks to address the feedback but it's done now.

I've decided against creating a new branch and committing my changes there (doing it the github way), since the repository that we are working on is not _very_ active and I think these extra commits might be counter productive.

I'm finding the process of review and rework interesting and educational, so if you think there is further rework needed, don't hesitate to request it.
